### ISSUE TITLE
caddy 2: reverse proxy to bitwarden?

### ISSUE THREAD

**Issue by sethgrid** [caddyserver/caddy#2943]
Hi Caddy team,
Thanks for caddy - I'm liking it so far and y'all have been very responsive. 
I'm trying to set up a reverse proxy to bitwarden but am getting 502 errors with not much more to go on. Curious how I should go about debugging it. 

The error log I get from caddy:
```
1.576867456328023e+09	error	http.log.error	read tcp 127.0.0.1:59660->127.0.0.1:5443: read: connection reset by peer	{"request": {"method": "GET", "uri": "/favicon.ico", "proto": "HTTP/2.0", "remote_addr": "13.52.196.42:58217", "host": "bitwarden.sethammons.com", "headers": {"Accept-Encoding": ["gzip, deflate, br"], "Pragma": ["no-cache"], "Cache-Control": ["no-cache"], "Te": ["trailers"], "User-Agent": ["Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:71.0) Gecko/20100101 Firefox/71.0"], "Accept": ["image/webp,*/*"], "Accept-Language": ["en-US,en;q=0.5"]}, "tls": {"resumed": true, "version": 772, "resumed": 4865, "proto": "h2", "proto_mutual": true, "server_name": "bitwarden.sethammons.com"}}, "status": 502, "err_id": "4t0bnerf1", "err_trace": "reverseproxy.(*Handler).ServeHTTP (reverseproxy.go:327)"}
```

I don't get any relevant data in sink log. My main site (sethammons.com) and my son's site (grzlybr.com) work fine, and his is reverse proxy too.

My caddy config:
```
{
    "logging": {
        "sink": {
            "writer": {
                "output": "file",
                "filename": "/var/log/caddy/sink.log"
            }
        },
        "logs": {
            "default": {
                "writer": {
                    "output": "file",
                    "filename": "/var/log/caddy/caddy.log"
                },
                "encoder": {
                    "format":"console"
                },
                "level": "info",
                "include": [],
                "exclude": []
            }
        }
    },
    "apps": {
        "http": {
            "servers": {
                "myserver": {
                    "listen": [
                        ":443"
                    ],
                    "routes": [
                         {
                            "match": [
                                {
                                    "host": [
                                        "bitwarden.sethammons.com"
                                    ]
                                }
                            ],
                            "handle": [
                                {
                                    "handler": "reverse_proxy",
                                    "upstreams": [
                                        {
                                            "dial": "localhost:5443",
                                            "max_requests": 1000
                                        }
                                    ]
                                }
                            ]
                        },

                        {
                            "match": [
                                {
                                    "host": [
                                        "sethammons.com"
                                    ]
                                }
                            ],
                            "handle": [
                                {
                                    "handler": "file_server",
                                    "root": "/home/seth/projects/me/site/public"
                                }
                            ]
                        },
                        {
                            "match": [
                                {
                                    "host": [
                                        "grzlybr.com"
                                    ]
                                }
                            ],
                            "handle": [
                                {
                                    "handler": "reverse_proxy",
                                    "upstreams": [
                                        {
                                            "dial": "localhost:1414",
                                            "max_requests": 1000
                                        }
                                    ]
                                }
                            ]
                        }
                    ]
                }
            }
        }
    }
}
```

I've reached out to the bitwarden team to see if there is anything else I can debug from their perspective and thought I'd ask here too. This feels like an issue with ssl forwarding as they expect https traffic. Oddly, bitwarden.sethammons.com:5080 (which I configured as the public port for bitwarden) works. But I can't reach bitwarden.sethammons.com:5443. 

When I do a tcpdump on 5443, I do see traffic back and forth:
```
18:52:27.553417 IP localhost.60000 > localhost.5443: Flags [F.], seq 1, ack 2, win 342, options [nop,nop,TS val 1993124217 ecr 1993124216], length 0
E..4..@.@../.........`.C).0M.......V.(.....
v..yv..x
18:52:27.553429 IP localhost.5443 > localhost.60000: Flags [.], ack 2, win 342, options [nop,nop,TS val 1993124217 ecr 1993124217], length 0
E..4S_@.@..b.........C.`....).0N...V.(.....
v..yv..y
```

It makes me believe that things are wired up correctly from the caddy side.

Thanks for any help y'all can offer. 
Cheers, --Seth

**Comment by mholt** [comment]
If BitWarden is expecting HTTPS, then you might want to configure the reverse proxy to use HTTPS on the upstream (between Caddy and BitWarden); by default it uses HTTP to the upstream.

You can enable HTTPS by specifying the transport in your reverse proxy config:

```
"transport": {
	"protocol": "http",
	"tls": {}
}
```

**Comment by sethgrid** [comment]
Ah, interesting. Do I need to fill anything in on the tls portion? Using the following config, the behavior is the same as before. 
```"routes": [
                         {
                            "match": [
                                {
                                    "host": [
                                        "bitwarden.sethammons.com"
                                    ]
                                }
                            ],
                            "handle": [
                                {
                                    "handler": "reverse_proxy",
                                    "transport": {
                                      "protocol": "http",
                                      "tls": {}
                                    },
                                    "upstreams": [
                                        {
                                            "dial": "localhost:5443",
                                            "max_requests": 1000
                                        }
                                    ]
                                }
                            ]
                        },
```

**Comment by mholt** [comment]
Nope, that should be all. Does it require client auth or something?

**Comment by sethgrid** [comment]
Ok, I got it all figured out. Your help to set up the transport was very solid. I had to tune some more configs on the bitwarden side. The way they are set up, I can't use caddy as ssl termination, it hast to pass through. So I re-installed bitwarden allowing it to set up its ssl cert, and updated all the internal configs to point to the fqdn:5443. And the way that bitwarden is set up, I needed caddy to proxy to bitwarden.sethammons.com:5443, not localhost:5443.

Thanks for all the help!
